talk/morelanguage/raii.tex

```diff
@@ -256,6 +284,113 @@
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[11]{weak\_ptr}
+  \begin{block}{weak\_ptr: a non-owning observer}
+    \small
+    \begin{itemize}
+    \item Sometimes want to observe resources without keeping them alive
+    \begin{itemize}
+      \item \mintinline{cpp}{shared_ptr}? Resource stays alive
+      \item Raw pointer? Risk of dangling pointer
+    \end{itemize}
+    \item The solution is to construct a \mintinline{cpp}{weak_ptr} from a shared pointer
+    \item To access the resource, convert the weak into a \mintinline{cpp}{shared_ptr}
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{}
+    \small
+    \begin{cppcode*}{}
+      std::shared_ptr<Cache>& getSharedCache();
+      std::weak_ptr<Cache> weakPtr{ getSharedCache() };
+      // ...
+      if (std::shared_ptr<Cache> cache = weakPtr.lock()) {
```

Comment:
OK, I think I got the example now. In that case, I think, you would not need a reference to the shared pointer and you may want to point out that the shared cache can be invalidated after taking the weak pointer:

```suggestion
      std::shared_ptr<Cache> getSharedCache();
      std::weak_ptr<Cache> weakPtr{ getSharedCache() };
      // ... shared cache may be invalidated here
      if (std::shared_ptr<Cache> cache = weakPtr.lock()) {
```